cli/tools/upgrade.rs

```diff
@@ -11,16 +13,103 @@ use deno_runtime::deno_fetch::reqwest::Client;
 use once_cell::sync::Lazy;
 use std::env;
 use std::fs;
+use std::io;
 use std::io::Write;
 use std::path::Path;
 use std::path::PathBuf;
 use std::process::Command;
+use std::time::Duration;
 
 static ARCHIVE_NAME: Lazy<String> =
   Lazy::new(|| format!("deno-{}.zip", env!("TARGET")));
 
 const RELEASE_URL: &str = "https://github.com/denoland/deno/releases";
 
+const UPGRADE_CHECK_INTERVAL: i64 = 12;
+
+const UPGRADE_CHECK_FETCH_DELAY: Duration = Duration::from_millis(500);
+
+pub fn check_for_upgrades() -> Option<String> {
+  if env::var("DENO_NO_UPDATE_CHECK").is_ok() {
+    return None;
+  }
+
+  let cache_dir = cache_dir()?;
+  let p = cache_dir.join("deno/latest.txt");
+  let content = match std::fs::read_to_string(&p) {
+    Ok(file) => file,
+    Err(err) if err.kind() == io::ErrorKind::NotFound => {
+      fs::create_dir_all(&cache_dir.join("deno")).ok()?;
+      if let Ok(c) = std::fs::read_to_string(&p) {
+        c
+      } else {
+        "".to_string()
+      }
+    }
+    Err(_) => "".to_string(),
+  };
+
+  let (mut last_checked, mut latest_version) = (None, None);
+  let split_content = content.split('!').collect::<Vec<_>>();
+  if split_content.len() == 2 {
+    last_checked = Some(split_content[0]);
+    latest_version = Some(split_content[1].to_owned());
+  }
+
+  if latest_version.is_none() || latest_version.as_ref().unwrap().is_empty() {
+    let mut last_checked_dt = None;
+
+    if let Some(last_checked) = last_checked {
+      if let Ok(dt) = chrono::DateTime::parse_from_rfc3339(last_checked) {
+        last_checked_dt = Some(dt.with_timezone(&chrono::Utc));
```

Comment:
Nitpick: Instead of having two separate options and mutating a variable, perhaps we could have a single function that retreives this information in a struct and returns a single option.